relay_handler.js

```diff
@@ -88,6 +118,257 @@ RelayHandler.prototype._handleRequest = function _handleRequest(req, buildRes) {
+    self.serviceName = '';
+    self.callerName = '';
+    self.timeout = 0;
+    self.timedOut = false;
+
+    self.boundExtendLogInfo = extendLogInfo;
+    self.boundOnIdentified = onIdentified;
+
+    function extendLogInfo(info) {
+        return self.extendLogInfo(info);
+    }
+
+    function onIdentified(err) {
+        if (err) {
+            self.onError(err);
+        } else {
+            self.onIdentified();
+        }
+    }
+}
+
+LazyRelayInReq.prototype.initRead =
+function initRead() {
+    var self = this;
+
+    var res = self.reqFrame.bodyRW.lazy.readTTL(self.reqFrame);
+    if (res.err) {
+        // TODO: wrap? protocol read error?
+        return res.err;
+    }
+    self.timeout = res.value;
+
+    res = self.reqFrame.bodyRW.lazy.readService(self.reqFrame);
+    if (res.err) {
+        // TODO: wrap? protocol read error?
+        return res.err;
+    }
+    self.serviceName = res.value;
+
+    // TODO: lazy read self.callerName
+
+    return null;
+};
+
+LazyRelayInReq.prototype.extendLogInfo =
+function extendLogInfo(info) {
+    var self = this;
+    // TODO: y u no request id? tracing id?
+    info.outRemoteAddr = self.outreq && self.outreq.remoteAddr;
+    info.inRemoteAddr = self.remoteAddr;
+
+    if (self.reqFrame) {
+        // TODO: maybe stash eagerly on inreq from constructor? maybe cache the
+        // rw through a memo on the frame (de-dupe read here and in service
+        // dispatch)?
+        var res = self.reqFrame.bodyRW.lazy.readService(self.reqFrame);
+        if (!res.err) {
+            info.serviceName = res.value;
+        }
+        // TODO: useful enough to implement a lazy reader for arg1?
+        // info.outArg1 = String(self.arg1);
+    }
+
+    return info;
+};
+
+LazyRelayInReq.prototype.logError =
+function relayRequestlogError(err, codeName) {
+    var self = this;
+    logError(self.conn.logger, err, codeName, function extendLogInfo(info) {
+        return self.extendLogInfo(info);
+    });
+};
+
+LazyRelayInReq.prototype.checkTimeout =
+function checkTimeout() {
+    var self = this;
+    if (!self.timedOut) {
+        var elapsed = self.conn.timers.now() - self.start;
+        if (elapsed > self.timeout) {
+            self.timedOut = true;
+            // TODO: cancel any outreq?
+            self.onError(errors.RequestTimeoutError({
+                id: self.id,
+                start: self.start,
+                elapsed: elapsed,
+                timeout: self.timeout
+            }));
+        }
+    }
+    return self.timedOut;
+};
+
+LazyRelayInReq.prototype.createOutRequest =
+function createOutRequest() {
+    var self = this;
+
+    if (self.outreq) {
+        self.conn.logger.warn('relay request already started', self.extendLogInfo({}));
+        return;
+    }
+
+    self.peer.waitForIdentified(self.boundOnIdentified);
+};
+
+LazyRelayInReq.prototype.onIdentified =
+function onIdentified(err) {
+    var self = this;
+
+    if (err) {
+        self.onError(err);
```

Comment:
good call, but that should be inside `#onError`, not just `#onIdentified`'s error branch; will add there